python_modules/dagster/dagster/_core/definitions/logical_version.py

```diff
@@ -0,0 +1,93 @@
+from __future__ import annotations
+
+from hashlib import sha256
+from typing import TYPE_CHECKING, AbstractSet, Mapping, Optional
+
+from typing_extensions import Final
+
+from dagster import _check as check
+
+if TYPE_CHECKING:
+    from dagster._core.definitions.events import AssetKey
+    from dagster._core.events.log import EventLogEntry
+    from dagster._core.instance import DagsterInstance
+
+class LogicalVersion:
+    """Class that represents a logical version for an asset.
+
+    Args:
+        value (str): An arbitrary string representing a logical version.
+    """
+
+    def __init__(self, value: str):
+        self.value = check.str_param(value, "value")
+
+    def __eq__(self, other: LogicalVersion):
+        return self.value == other.value
+
+
+DEFAULT_LOGICAL_VERSION: Final[LogicalVersion] = LogicalVersion("DEFAULT")
+
+
+def get_logical_version_from_inputs(
+    dependency_keys: AbstractSet[AssetKey],
+    op_version: str,
+    key_to_is_source_map: Mapping[AssetKey, bool],
+    instance: DagsterInstance,
+    logical_version_overrides: Optional[Mapping[AssetKey, LogicalVersion]] = None,
+) -> LogicalVersion:
+    logical_version_overrides = check.opt_mapping_param(
+        logical_version_overrides, "logical_version_overrides"
+    )
+    ordered_dependency_keys = sorted(dependency_keys, key=str)
+    dependency_logical_versions = [
+        logical_version_overrides[k]
+        if k in logical_version_overrides
+        else get_most_recent_logical_version(k, key_to_is_source_map[k], instance)
+        for k in ordered_dependency_keys
+    ]
+    all_inputs = [op_version, *dependency_logical_versions]
+    hash_sig = sha256()
+    hash_sig.update(bytearray("".join(all_inputs), "utf8"))
+    return LogicalVersion(hash_sig.hexdigest())
+
+
+def get_most_recent_logical_version(
```

Comment:
I would expect this to live inside the instance.

Reply:
also, can we do some caching or something to avoid fetching the same event multiple times within a single step?